cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp

```diff
@@ -0,0 +1,258 @@
+            ROS_WARN_STREAM("JointName: "<<joint.getName());
+            ROS_WARN_STREAM("JointType: "<<joint.getTypeName());
+        }
+        ROS_WARN("~~~~~~~~~~~~~ DEBUG CHAIN ~~~~~~~~~~~~~");
+
+        p_fksolver_vel_ = new KDL::ChainFkSolverVel_recursive(chain_);
+        p_jnt2jac_ = new KDL::ChainJntToJacSolver(chain_);
+
+        /// initialize ROS interfaces
+        // joint_states_sub_ = nh_.subscribe("joint_states", 1, &DebugFkVelRecursive::jointstatesCallback, this);  // analyzing controller_state instead
+        controller_state_sub_ = nh_.subscribe("joint_trajectory_controller/state", 1, &DebugFkVelRecursive::controllerstateCallback, this);
+        manipulability_pub_ = nh_.advertise<std_msgs::Float64> ("debug/manipulability", 1);
+        twist_current_pub_ = nh_.advertise<geometry_msgs::Twist> ("debug/twist_current", 1);
+        twist_magnitude_desired_pub_ = nh_.advertise<sensor_msgs::JointState> ("debug/twist_magnitude/desired", 1);
+        twist_magnitude_actual_pub_ = nh_.advertise<sensor_msgs::JointState> ("debug/twist_magnitude/actual", 1);
+
+        return 0;
+    }
+
+    void jointstatesCallback(const sensor_msgs::JointState::ConstPtr& msg)
+    {
+        KDL::JntArray q = KDL::JntArray(chain_.getNrOfJoints());
+        KDL::JntArray q_dot = KDL::JntArray(chain_.getNrOfJoints());
+
+        /// extract q, q_dot from JointStates
+        for (unsigned int i = 0; i < chain_.getNrOfSegments(); i++)
+        {
+            KDL::Joint joint = chain_.getSegment(i).getJoint();
+            std::string joint_name = joint.getName();
+            if (joint.getType() == KDL::Joint::None)
+            {
+                ROS_DEBUG_STREAM("Skip fixed Joint '" << joint_name);
+                continue;
+            }
+            else
+            {
+                if (std::find(msg->name.begin(), msg->name.end(), joint_name) != msg->name.end())
+                {
+                    unsigned int index = std::distance(msg->name.begin(), std::find(msg->name.begin(), msg->name.end(), joint_name));
+                    q(index) = msg->position[index];
+                    q_dot(index) = msg->velocity[index];
+                }
+                else
+                {
+                    ROS_ERROR_STREAM("Joint '" << joint_name << "' not found in JointStates");
+                    return;
+                }
+            }
+        }
+
+        /// compute current twist
+        std::vector< KDL::FrameVel > v_FrameVel;
+        KDL::JntArrayVel jntArrayVel = KDL::JntArrayVel(q, q_dot);
+        v_FrameVel.resize(chain_.getNrOfSegments());
+        if (p_fksolver_vel_->JntToCart(jntArrayVel, v_FrameVel, chain_.getNrOfSegments()) >= 0)
+        {
+            //last entry is twist of tip_link
+            geometry_msgs::Twist twist_msg;
+            tf::twistKDLToMsg(v_FrameVel.back().GetTwist(), twist_msg);
+            twist_current_pub_.publish(twist_msg);
+
+            //recursively calculate FrameVel magnitudes
```

Comment:
I don't see what is recursive here. Code just iterates over the joints right?

Or do I understand wrong?

Reply:
Ah, the FK solver works recursively, it's a https://docs.ros.org/en/indigo/api/orocos_kdl/html/classKDL_1_1ChainFkSolverVel__recursive.html